File: ROS-Services-and-Actions/service_action_examples_package/src/square_client.cpp
#include <ros/ros.h>
#include <service_action_examples_package/Square.h>

int main(int argc, char **argv)
{
    // Initialize the ROS node
    ros::init(argc, argv, "square_client");
    ros::NodeHandle nh;
    
    ROS_INFO("spp square_client");

    // Create a service client for 'calculate_square'
    ros::ServiceClient client = nh.serviceClient<service_action_examples_package::Square>("calculate_square");

    if (argc != 2)
    {
        ROS_INFO("Usage: square_client [number]");
        return 1;
    }

    // Create a request object and set the number
    service_action_examples_package::Square srv;
    srv.request.number = atoi(argv[1]);

    // Call the service and check the result
    if (client.call(srv))
    {
        ROS_INFO("The square of %d is %d", srv.request.number, srv.response.square);
    }
    else
    {
        ROS_ERROR("Failed tooooo call service calculate_square");
        return 1;
    }

    return 0;
}

File: ROS Services And Actions/service_action_examples_package/scripts/square_client.py
#!/usr/bin/env python3

import sys
import rospy
from service_action_examples_package.srv import Square

def square_client(number):
    # Wait for the 'calculate_square' service to become available
    rospy.wait_for_service('calculate_square')
    try:
        # Create a handle to the service
        calculate_square = rospy.ServiceProxy('calculate_square', Square)
        # Call the service with the provided number
        response = calculate_square(number)
        # Return the result from the service response
        return response.square
    except rospy.ServiceException as e:
        # Print error if the service call fails
        rospy.logerr("Service call failed: %s" % e)

if __name__ == "__main__":
    # Check if a number was provided as a command-line argument
    if len(sys.argv) == 2:
        number = int(sys.argv[1])
    else:
        print("Usage: square_client.py [number]")
        sys.exit(1)
    
    # Print the number being requested
    print("Requesting the square of %d" % number)
    # Call the service and print the result
    print("Square: %d" % square_client(number))
